File: backend/src/bus/subjects.ts
export const SUBJECTS = {
  // AI RPC (request-reply через JetStream, чтобы воркер не брал новое пока занят)
  aiDetectLanguage: 'ai.rpc.detect_language',
  aiSearchQueries: 'ai.rpc.search_queries',
  aiRankLyrics: 'ai.rpc.rank_lyrics',
  aiTranscribe: 'ai.rpc.transcribe',
  aiEncodeTextMulan: 'ai.rpc.encode_text_mulan',
  aiLtrScore: 'ai.rpc.ltr_score',

  // JetStream work queues (backend → worker, durable)
  indexAudio: 'index.audio.new',
  embedLyrics: 'embed.lyrics.new',
  trainCollab: 'train.collab.new',
  trainLtr: 'train.ltr.new',

  // JetStream fan-out (worker → backend notify о завершении)
  doneIndexAudio: 'done.index_audio',
  doneEmbedLyrics: 'done.embed_lyrics',
  doneTrainCollab: 'done.train_collab',
  doneTrainLtr: 'done.train_ltr',

  // JetStream fan-out (streaming → backend notify об аплоаде в storage)
  storageTrackUploaded: 'storage.track_uploaded',
} as const;

export const STREAMS = {
  // AI RPC work queue — короткий TTL, чтобы устаревшие запросы не висели в стриме.
  aiRpc: {
    name: 'AI_RPC',
    subjects: ['ai.rpc.>'],
  },
  // Backend → worker, durable consumers на стороне воркера
  indexAudio: {
    name: 'INDEX_AUDIO',
    subjects: ['index.audio.>'],
  },
  embedLyrics: {
    name: 'EMBED_LYRICS',
    subjects: ['embed.lyrics.>'],
  },
  trainCollab: {
    name: 'TRAIN_COLLAB',
    subjects: ['train.collab.>'],
  },
  trainLtr: {
    name: 'TRAIN_LTR',
    subjects: ['train.ltr.>'],
  },
  // Worker → backend notifications
  done: {
    name: 'PIPELINE_DONE',
    subjects: ['done.>'],
  },
  // Streaming → backend notifications (storage.track_uploaded и т.п.)
  storageEvents: {
    name: 'STORAGE_EVENTS',
    subjects: ['storage.>'],
  },
} as const;

export const QUEUE_GROUPS = {
  aiWorkers: 'ai-workers',
} as const;

File: backend/src/bus/subjects.rs
pub mod subjects {
    pub const AI_DETECT_LANGUAGE: &str = "ai.rpc.detect_language";
    pub const AI_SEARCH_QUERIES: &str = "ai.rpc.search_queries";
    pub const AI_RANK_LYRICS: &str = "ai.rpc.rank_lyrics";
    pub const AI_TRANSCRIBE: &str = "ai.rpc.transcribe";
    pub const AI_ENCODE_TEXT_MULAN: &str = "ai.rpc.encode_text_mulan";
    pub const AI_LTR_SCORE: &str = "ai.rpc.ltr_score";
    pub const AI_RESOLVE_ARTIST: &str = "ai.rpc.resolve_artist";
    pub const AI_VERIFY_EXISTENCE: &str = "ai.rpc.verify_existence";
    pub const AI_MATCH_TRACK: &str = "ai.rpc.match_track";
    pub const AI_TWO_TOWER_SCORE: &str = "ai.rpc.two_tower_score";
    pub const AI_SEQUENTIAL_PREDICT: &str = "ai.rpc.sequential_predict";
    pub const AI_QUALITY_SCORE: &str = "ai.rpc.quality_score";

    pub const INDEX_AUDIO: &str = "index.audio.new";
    pub const EMBED_LYRICS: &str = "embed.lyrics.new";
    pub const TRAIN_COLLAB: &str = "train.collab.new";
    pub const TRAIN_LTR: &str = "train.ltr.new";
    pub const TRAIN_TWO_TOWER: &str = "train.two_tower.new";
    pub const TRAIN_SEQUENTIAL: &str = "train.sequential.new";
    pub const TRAIN_QUALITY: &str = "train.quality.new";

    pub const ENRICH_TRACK: &str = "enrich.track.new";

    pub const DONE_INDEX_AUDIO: &str = "done.index_audio";
    pub const DONE_EMBED_LYRICS: &str = "done.embed_lyrics";

    pub const STORAGE_TRACK_UPLOADED: &str = "storage.track_uploaded";
}

pub struct StreamCfg {
    pub name: &'static str,
    pub subjects: &'static [&'static str],
}

pub mod streams {
    use super::StreamCfg;

    pub const AI_RPC: StreamCfg = StreamCfg {
        name: "AI_RPC",
        subjects: &["ai.rpc.>"],
    };
    pub const INDEX_AUDIO: StreamCfg = StreamCfg {
        name: "INDEX_AUDIO",
        subjects: &["index.audio.>"],
    };
    pub const EMBED_LYRICS: StreamCfg = StreamCfg {
        name: "EMBED_LYRICS",
        subjects: &["embed.lyrics.>"],
    };
    pub const TRAIN_COLLAB: StreamCfg = StreamCfg {
        name: "TRAIN_COLLAB",
        subjects: &["train.collab.>"],
    };
    pub const TRAIN_LTR: StreamCfg = StreamCfg {
        name: "TRAIN_LTR",
        subjects: &["train.ltr.>"],
    };
    pub const TRAIN_TWO_TOWER: StreamCfg = StreamCfg {
        name: "TRAIN_TWO_TOWER",
        subjects: &["train.two_tower.>"],
    };
    pub const TRAIN_SEQUENTIAL: StreamCfg = StreamCfg {
        name: "TRAIN_SEQUENTIAL",
        subjects: &["train.sequential.>"],
    };
    pub const TRAIN_QUALITY: StreamCfg = StreamCfg {
        name: "TRAIN_QUALITY",
        subjects: &["train.quality.>"],
    };
    pub const ENRICH: StreamCfg = StreamCfg {
        name: "ENRICH",
        subjects: &["enrich.>"],
    };
    pub const DONE: StreamCfg = StreamCfg {
        name: "PIPELINE_DONE",
        subjects: &["done.>"],
    };
    pub const STORAGE_EVENTS: StreamCfg = StreamCfg {
        name: "STORAGE_EVENTS",
        subjects: &["storage.>"],
    };
}
